---
title: "Visualize GWAS output"
# author: Yanyu Liang
date: "`r format(Sys.time(), '%b %d, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: cosmo
    highlight: textmate
    number_sections: true
---

```{r setup}
library(data.table)
library(dplyr)
library(ggplot2)
options(datatable.fread.datatable = F)
parse_allele = function(str) {
  unlist(lapply(strsplit(str, '\\",\\"'), function(x) {
    a1 = stringr::str_remove(x[1], '\\[\\"')
    a2 = stringr::str_remove(x[2], '\\"\\]')
    return(paste0(a1, ':', a2))
  }))
}
get_snp = function(locus, alleles) {
  paste0(locus, ':', parse_allele(alleles))
}
```

# About

See the script that generated GWAS results at [here](https://github.com/liangyy/toy-example-of-hail-using-1kg/blob/master/notebook/run_multi_trait_gwas.ipynb).

# Result

Load all subsets and all traits.
Also, we load in true effects.

```{r load}
df_gwas = list()
for(i in 0 : 1) {
  for(j in 0 : 11) {
    tmp = fread(paste0('output/gwas_subset_', i, '_x_trait_', j, '.tsv'), sep = '\t')
    df_gwas[[length(df_gwas) + 1]] = tmp %>% mutate(subset = i, trait = paste0('trait_', j))
  }
}
df_gwas = do.call(rbind, df_gwas)
df_true = fread('output/snp_effect_size.tsv', sep = '\t')
df_true = df_true %>% mutate(variant = get_snp(locus, alleles))
```

## QQ plot

And do QQ-plot aggregating across all results.

```{r qq}
source('https://gist.githubusercontent.com/liangyy/605437935751bb77b1739666b18517bf/raw/85b1e83d1cbf7c8b5671030e927000342ed16f97/my_qqplot_by_group.R')
my_qqplot_by_group(df_gwas$pval, factor(df_gwas$trait)) + facet_wrap(~grp)
```

## Compare estimated effect size with truth

```{r compare, fig.height=12, fig.width=3}
tmp = df_gwas %>% inner_join(
  df_true %>% select(-locus, -alleles) %>% melt(id.vars = 'variant') %>% rename(true_effect = value),
  by = c('variant', 'trait' = 'variable')
)
tmp %>% ggplot() + geom_point(aes(x = true_effect, y = beta), alpha = .2) + facet_grid(trait~subset) + geom_abline(intercept = 0, slope = 1)
```